### PR TITLE
[SPARK-41271][SQL] Support parameterized SQL queries by `sql()`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to extend SparkSession API and override the `sql` method by:
```scala
  def sql(sqlText: String, args: Map[String, String]): DataFrame
```
which accepts a map with:
- keys are parameters names,
- values are SQL literal values.

And the first argument `sqlText` might have named parameters in the positions of constants like literal values.

For example:
```scala
  spark.sql(
    sqlText = "SELECT * FROM tbl WHERE date > :startDate LIMIT :maxRows",
    args = Map(
      "startDate" -> "DATE'2022-12-01'",
      "maxRows" -> "100"))
```
The new `sql()` method parses the input SQL statement and provided parameter values, and replaces the named parameters by the literal values. And then it eagerly runs DDL/DML commands, but not for SELECT queries.

Closes #38712

### Why are the changes needed?
1. To improve user experience with Spark SQL via
    - Using Spark as remote service (microservice).
    - Write SQL code that will power reports, dashboards, charts and other data presentation solutions that need to account for criteria modifiable by users through an interface.
    - Build a generic integration layer based on the SQL API. The goal is to expose managed data to a wide application ecosystem with a microservice architecture. It is only natural in such a setup to ask for modular and reusable SQL code, that can be executed repeatedly with different parameter values.

2. To achieve feature parity with other systems that support named parameters:
    - Redshift: https://docs.aws.amazon.com/redshift/latest/mgmt/data-api.html#data-api-calling
    - BigQuery: https://cloud.google.com/bigquery/docs/parameterized-queries#api
    - MS DBSQL: https://learn.microsoft.com/en-us/azure/databricks/sql/user/queries/query-parameters

### Does this PR introduce _any_ user-facing change?
No, this is an extension of the existing APIs.

### How was this patch tested?
By running new tests:
```
$ build/sbt "core/testOnly *SparkThrowableSuite"
$ build/sbt "test:testOnly *PlanParserSuite"
$ build/sbt "test:testOnly *AnalysisSuite"
$ build/sbt "test:testOnly *ParametersSuite"

```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @entong Could you take a look at this PR, please.

**Comment by xkrogen** [comment]
What is the relationship between this PR and #38712? Why do we have two PRs?

If this PR is superceding #38712, can we continue discussion here on which identifier to use based on [my last comment on the old PR](https://github.com/apache/spark/pull/38712#issuecomment-1334227405)?

**Comment by MaxGekk** [comment]
> What is the relationship between this PR and https://github.com/apache/spark/pull/38712? Why do we have two PRs?

@xkrogen The previous one was a draft PR. This one supersedes it. 

> can we continue discussion here ...

Yep.

> please let me know if you have a better source ...

As I can understand the SQL standard, it doesn't restrict identifiers in our case. The standard requires the `:` marker in another cases:
1. Names of host parameters in [externally-invoked procedure](https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#_13_3_externally_invoked_procedure)

https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#host-parameter-name

2.  Embedded variable names

https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#embedded-variable-name

cc @srielau

**Comment by xkrogen** [comment]
I managed to find a better SQL standard reference in the form of _SQL: The Complete Reference_ (2003), which has an entire chapter devoted to Dynamic SQL (beginning from page 547). You are correct regarding colons; those are indeed _host variables_. The standard for `PREPARE`/`EXECUTE` refers to positional arguments using `?`; this makes sense given how widely support that syntax is ([as mentioned in my last comment](https://github.com/apache/spark/pull/38712#issuecomment-1334227405)).

However, I think it is safe to make an argument that what you are trying to add here is not, in fact, dynamic SQL, but rather host variables. We have no notion of precompiled statements with `PREPARE`/`EXECUTE`, which is the hallmark of dynamic SQL. Rather we have a query with a few "blanks" to be filled in, and we want to provide those blanks from the host execution environment. That is exactly what a host variable is for, right? You happen to be proposing an API (in Scala) that provides the values of the variables as map, instead of pulling them directory from the JVM execution environment, but from a SQL syntax perspective they are identical.

i.e. you propose this example (syntax updated to `:`):
```scala
  spark.sql(
    sqlText = "SELECT * FROM tbl WHERE date > :startDate LIMIT :maxRows",
    args = Map(
      "startDate" -> "DATE'2022-12-01'",
      "maxRows" -> "100"))
```
This SQL syntax is fully compatible with the standard for host variables, it's just that the Scala interface is slightly different. This would _not_ preclude us from supporting the following in the future using the exact same SQL syntax:
```scala
  val startDate = "DATE'2022-12-01'"
  val maxRows = "100"
  spark.sql(sqlText = "SELECT * FROM tbl WHERE date > :startDate LIMIT :maxRows")
```
Both are fully compatible; one is explicit and one is implicit.

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaSparkSessionSuite.java [hunk 1]
```diff
@@ -54,4 +55,19 @@ public void config() {
       Assert.assertEquals(spark.conf().get(e.getKey()), e.getValue().toString());
     }
   }
+
+  @Test
+  public void sqlParameters() {
+    spark = SparkSession.builder().master("local[*]").appName("testing").getOrCreate();
+    Map params = new HashMap();
+    params.put("_i1", "INTERVAL '1-1' YEAR TO MONTH");
+    params.put("p2", "'abc'");
```
Comment (entong) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaSparkSessionSuite.java:64]: Could you add a test case where the string literal contains a double quote?

Comment (MaxGekk) [reply sql/core/src/test/java/test/org/apache/spark/sql/JavaSparkSessionSuite.java]: added

**Comment by MaxGekk** [comment]
> I could just as easily say that we should choose : to make it easier to migrate from Redshift ...

How about to support both `@` and `:`? In that way, we cover and encourage more users to migrate on to Spark SQL.

**Comment by xkrogen** [comment]
Regarding supporting two syntaxes vs. standardizing on one, I don't have much of an opinion, but I would be curious about how Spark has handled such decisions in the past when there are competing dialects. @cloud-fan, any thoughts here?

**Comment by MaxGekk** [comment]
@xkrogen After offline discussion with @cloud-fan @srielau, we decided to change the parameter marker to `:` in the PR.

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -795,6 +795,11 @@
       }
     }
   },
+  "INVALID_SQL_ARG" : {
+    "message" : [
+      "The argument <name> of `sql()` is invalid. Consider to replace it by a literal value."
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:800]:
```suggestion
      "The argument <name> of `sql()` is invalid. Consider to replace it by a SQL literal statement."
```

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -1130,6 +1135,11 @@
       "Unable to convert SQL type <toType> to Protobuf type <protobufType>."
     ]
   },
+  "UNBOUND_PARAMETER" : {
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:1138]:
```suggestion
  "UNBOUND_QUERY_PARAMETER" : {
```

File: core/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -795,6 +795,11 @@
       }
     }
   },
+  "INVALID_SQL_ARG" : {
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:798]:
```suggestion
  "INVALID_QUERY_PARAMETER" : {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/parameters.scala [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.spark.sql.errors.{QueryErrorsBase, QueryExecutionErrors}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{DataType, NullType}
+
+/**
+ * The expression represents a named parameter that should be replaces by a foldable expression.
+ *
+ * @param name The identifier of the parameter without the marker.
+ */
+case class Parameter(name: String) extends LeafExpression {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/parameters.scala:34]: This should be unresolved.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 6]
```diff
@@ -2245,6 +2245,25 @@ class DatasetSuite extends QueryTest
       assert(parquetFiles.size === 10)
     }
   }
+
+  test("SPARK-41271: bind parameters") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:2249]: can we have a new test suite for it?

**Comment by MaxGekk** [comment]
After switching from `@` to `:`, I got many parsing errors (see [link](https://github.com/MaxGekk/spark/runs/9964396959)):
```
SchemaPruningSuite.prune schema by the requested fields
org.apache.spark.sql.catalyst.parser.ParseException: 
[PARSE_SYNTAX_ERROR] Syntax error at or near ':int'(line 1, pos 10)

== SQL ==
a struct<a:int, b:int>, b int
----------^^^
```

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4 [hunk 7]
```diff
@@ -930,6 +935,7 @@ primaryExpression
 
 constant
     : NULL                                                                                     #nullLiteral
+    | {parameters_enabled}? ':' identifier                                                     #parameterLiteral
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:938]: let's match the token `COLON`

Comment (MaxGekk) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4]: done

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala [hunk 8]
```diff
@@ -667,6 +667,34 @@ class QueryCompilationErrorsSuite
...
+      },
+      errorClass = "UNBOUND_PARAMETER",
+      parameters = Map("name" -> "def"),
+      context = ExpectedContext(
+        fragment = ":def",
+        start = 13,
+        stop = 16))
+  }
+
+  test("INVALID_SQL_ARG - SPARK-41271: non-literal argument of `sql()`") {
```
[the remark this answers is not in the document]
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala]: Moved to `ParametersSuite`

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala (hunk 6 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Created this `ParametersSuite`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/parameters.scala (hunk 5 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/parameters.scala:34]: I will add the override:
```scala
override lazy val resolved: Boolean = false
```

**Comment by MaxGekk** [comment]
@cloud-fan Could you look at the PR one more time, please.

**Comment by srielau** [comment]
@MaxGekk Could you sketch out how this API could be extended to INOUT and OUT parameters in the future?

**Comment by MaxGekk** [comment]
> @MaxGekk Could you sketch out how this API could be extended to INOUT and OUT parameters in the future?

@srielau This feature precedences INOUT/OUT parameters. Apparently, the `sql()` method will substitute only parameters specified in `args`, and will leave others as INOUT params. Till the INOUT feature is not implemented in OSS, Spark fails w/ `UNBOUND_SQL_PARAMETER` when it faces to an param w/o provided argument.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4 [hunk 9]
```diff
@@ -40,6 +40,11 @@ options { tokenVocab = SqlBaseLexer; }
    * When true, double quoted literals are identifiers rather than STRINGs.
    */
   public boolean double_quoted_identifiers = false;
+
+  /**
+   * When true, identifiers that begin from `:` are considered as named parameters.
+   */
+  public boolean parameters_enabled = false;
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:47]: This is a new SQL syntax, it won't break any existing queries. Do we really need a feature flag?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/parameters.scala [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+  override lazy val resolved: Boolean = false
+  override def dataType: DataType = NullType
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/parameters.scala:38]: we can fail here, similar to what `UnresolvedAttribute.dataType` does

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/parameters.scala [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  override def nullable: Boolean = true
+
+  final override val nodePatterns: Seq[TreePattern] = Seq(PARAMETER)
+
+  private def unboundError(): Nothing = {
+    throw SparkException.internalError(
+      s"The parameter `$name` must be bound at the analysis phase.")
+  }
+
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = unboundError()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/parameters.scala:48]: We can just extend `Unevaluable`

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaSparkSessionSuite.java [hunk 12]
```diff
@@ -54,4 +55,19 @@ public void config() {
       Assert.assertEquals(spark.conf().get(e.getKey()), e.getValue().toString());
     }
   }
+
+  @Test
+  public void sqlParameters() {
```
Comment (cloud-fan) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaSparkSessionSuite.java:60]: do we need a java test? Adding a overload that takes a java map is very trivial...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/parameters.scala (hunk 10 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/parameters.scala:38]: Thank you. I will throw an internal error. I think it is more appropriate here then a sub-class of `AnalysisException`.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4 (hunk 9 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:47]: I have removed the SQL config.

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaSparkSessionSuite.java (hunk 12 above)
Comment (MaxGekk) [reply sql/core/src/test/java/test/org/apache/spark/sql/JavaSparkSessionSuite.java]: I removed it.

**Comment by MaxGekk** [comment]
@cloud-fan @entong Could you review this PR one more time, please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/parameters.scala [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+import org.apache.spark.SparkException
+import org.apache.spark.sql.catalyst.analysis.AnalysisErrorAt
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.trees.TreePattern.{PARAMETER, TreePattern}
+import org.apache.spark.sql.errors.QueryErrorsBase
+import org.apache.spark.sql.types.DataType
+
+/**
+ * The expression represents a named parameter that should be replaces by a literal.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/parameters.scala:28]:
```suggestion
 * The expression represents a named parameter that should be replaced by a literal.
```

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
Merging to master. The last commit is minor one.
Thank you, @cloud-fan @xkrogen @entong @srielau for review.

File: core/src/main/resources/error/error-classes.json [hunk 14]
```diff
@@ -795,6 +795,11 @@
...
+    "message" : [
+      "The argument <name> of `sql()` is invalid. Consider to replace it by a SQL literal statement."
```
Comment (srielau) [inline core/src/main/resources/error/error-classes.json:800]: Can we more explicit about why it is invalid?
"The argument <name> of `sql()` is not a literal." What is a "SQL Literal statement"?

File: core/src/main/resources/error/error-classes.json [hunk 15]
```diff
@@ -1130,6 +1135,11 @@
       "Unable to convert SQL type <toType> to Protobuf type <protobufType>."
     ]
   },
+  "UNBOUND_SQL_PARAMETER" : {
+    "message" : [
+      "Found the unbound parameter: <name>. Please, fix `args` and provide a mapping of the parameter to a SQL literal statement."
```
Comment (srielau) [inline core/src/main/resources/error/error-classes.json:1140]: Won't this same error be used fro all other APIs (JDBC, SQL (when we support). So we may not want to refer to `args`.

File: core/src/main/resources/error/error-classes.json (hunk 14 above)
Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:800]: > What is a "SQL Literal statement"?

Any SQL statement that can produce a literal, see https://spark.apache.org/docs/latest/sql-ref-literals.html

File: core/src/main/resources/error/error-classes.json (hunk 15 above)
Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:1140]: > So we may not want to refer to args.

This is a premature generalisation.  Let's make it more generic when we will need that.

File: core/src/main/resources/error/error-classes.json (hunk 14 above)
Comment (srielau) [reply core/src/main/resources/error/error-classes.json:800]: That's an expression then. Statements are top level. (Like SELECT, UPDATE, CRAETE, SET).

Comment (cloud-fan) [reply core/src/main/resources/error/error-classes.json:800]: how about just say `SQL literal`?

Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:800]: I will change this in https://github.com/apache/spark/pull/39183

**Pull request closed without merging**